vm_supervisor/network/hostnetwork.py

```diff
@@ -1,26 +1,144 @@
+
+
+def get_ipv6_forwarding_state() -> int:
+    """Reads the current IPv6 forwarding setting from the host, converts it to int and returns it"""
+    return _read_proc_file_value(Path("/proc/sys/net/ipv6/conf/all/forwarding"))
+
+
+class IPv6Allocator(Protocol):
+    def allocate_vm_ipv6_subnet(self, vm_id: int, vm_hash: ItemHash) -> IPv6Network:
+        ...
+
+
+class StaticIPv6Allocator(IPv6Allocator):
+    """
+    Static IPv6 allocator.
+    Computes IPv6 addresses based on the machine type and VM hash. The algorithm works
+    as follows:
+
+    | Component | CRN /64 range | VM type | Item hash prefix | Instance range |
+    |-----------|---------------|---------|------------------|----------------|
+    | Length    | 64 bits       | 16 bits | 44 bits          | 4 bits         |
+    """
+
+    def __init__(self, ipv6_range: IPv6Network, subnet_prefix: int):
+        if ipv6_range.prefixlen != 64:
+            raise ValueError(
+                "The static IP address allocation scheme requires a /64 subnet"
+            )
+        if subnet_prefix < 124:
+            raise ValueError("The IPv6 subnet prefix cannot be larger than /124.")
+
+        self.ipv6_range = ipv6_range
+        self.subnet_prefix = subnet_prefix
+
+    def allocate_vm_ipv6_subnet(self, vm_id: int, vm_hash: ItemHash) -> IPv6Network:
+        ipv6_elems = self.ipv6_range.exploded.split(":")[:4]
+        ipv6_elems += ["1"]  # Magic number for microVMs
+        ipv6_elems += [vm_hash[0:4], vm_hash[4:8], vm_hash[8:11] + "0"]
+
+        return IPv6Network(":".join(ipv6_elems) + "/124")
+
+
+class DynamicIPv6Allocator(IPv6Allocator):
+    """
+    A dynamic allocator, for testing purposes.
+    This allocator slices the input IPv6 address range in subnets of the same size
+    and iterates through them. The first subnet is assumed to be reserved for use by the host,
+    as we use this allocator in situations where the address range is small and the host
+    subnet cannot be isolated from the VM subnets (ex: /124 network on Digital Ocean for the CI).
+    """
+
+    def __init__(self, ipv6_range: IPv6Network, subnet_prefix: int):
+        self.ipv6_range = ipv6_range
+        self.vm_subnet_prefix = subnet_prefix
+
+        self.subnets_generator = ipv6_range.subnets(new_prefix=subnet_prefix)
+        # Assume the first subnet is reserved for the host
+        _ = next(self.subnets_generator)
+
+    def allocate_vm_ipv6_subnet(self, vm_id: int, vm_hash: ItemHash) -> IPv6Network:
+        return next(self.subnets_generator)
+
+
+def make_ipv6_allocator(
+    allocation_policy: IPv6AllocationPolicy, address_pool: str, subnet_prefix: int
+) -> IPv6Allocator:
+    allocator_cls = (
+        StaticIPv6Allocator
+        if allocation_policy == IPv6AllocationPolicy.static
+        else DynamicIPv6Allocator
+    )
+    return allocator_cls(
+        ipv6_range=IPv6Network(address_pool), subnet_prefix=subnet_prefix
+    )
```

Comment:
Storing the class into a variable here does not seem to make the code simpler to read...

```suggestion
    if allocation_policy == IPv6AllocationPolicy.static:
        return StaticIPv6Allocator(ipv6_range=IPv6Network(address_pool), subnet_prefix=subnet_prefix)
    else:
        return DynamicIPv6Allocator(ipv6_range=IPv6Network(address_pool), subnet_prefix=subnet_prefix)
```